Assigning Joystick to BBC Keyboard

monkeyman79 opened the pull request:
- Assigning host joystick axes and buttons to BBC keyboard (up to 2 joysticks or gamepads separately).
- New human readable file format for joystick mapping (see Documents/JoystickMap.txt)
- Option to automatically load image-specific joystick mapping file (same as disk name, but with .jmap extension)
- Fixed displaying keys assigned to Shift in User Mapping dialog
- Added "unassign" to keyboard and joystick mapping dialog
- Joystick mapping table is separate from key mapping table
- Mapping joysticks is done separately from keyboard (but using the same dialog code)
- Highlight assigned keys in joystick mapping dialog

This is redo of PR #44

monkeyman79 commented:
It's been waiting a long time, but finally it's 100% done. Joystick mapping is completely separate from user keyboard mapping, loaded, saved and configured separately. Grab it while it's hot.

Code reviews are welcome.

Don't hesitate to change or rewrite my additions to help pages - I'm not a native speaker and I may have made some stylistic or other mistakes.

monkeyman79 commented on Documents/JoystickMap.txt:
```diff
@@ -0,0 +1,70 @@
+
+# Special BBC key names are:
+# LEFT, RIGHT, UP, DOWN, BREAK, COPY, DELETE, CAPS-LOCK, TAB, CTRL
+# SPACE, RETURN, ESCACE, SHIFT, SHIFT-LOCK
```
Typo: ESCAPE

chrisn commented on Src/beebwin.cpp:
```diff
@@ -1248,13 +1294,179 @@ void BeebWin::ScaleJoystick(unsigned int x, unsigned int y)
+}
+
 /****************************************************************************/
 void BeebWin::ResetJoystick(void)
```
Should we remove this function, if it doesn't do anything?

monkeyman79 replied:
Yes

chrisn commented on Src/beebwin.cpp:
```diff
@@ -1609,11 +1821,17 @@ LRESULT CALLBACK WndProc(HWND hWnd,     // window handle
 			break;
 
 		case MM_JOY1BUTTONDOWN:
 		case MM_JOY1BUTTONUP:
```
Do we also need to handle `MM_JOY2BUTTONDOWN` and `MM_JOY2BUTTONUP`?

monkeyman79 replied:
Pushing or releasing joystick button generates `MM_JOY2MOVE` in addition to `MM_JOY2BUTTONDOWN` or `MM_JOY2BUTTONUP`. There's no need to handle those events separately.

chrisn commented on Src/beebwin.h:
```diff
@@ -302,15 +345,27 @@ class BeebWin {
 	int		m_FPSTarget;
 	bool		m_JoystickCaptured;
 	JOYCAPS		m_JoystickCaps;
+	unsigned int	m_Joystick1Deadband;
```
I suggest using an array of structs here, e.g:

```c++
struct JoystickState {
    JOYCAPS Caps;
    unsigned int Deadband;
    int PrevAxes;
    int PrevBtns;
    bool Captured;
};
```
and then
```c++
JoystickState m_JoystickState[2];
```

This would simplify the code in beebwin.cpp where you have:

```c++
int& prevBtns = (joyId == 1) ? m_Joystick2PrevBtns : m_Joystick1PrevBtns;
```

monkeyman79 replied:
Agreed

chrisn commented on Src/beebwin.h:
```diff
@@ -392,6 +447,7 @@ class BeebWin {
 	bool		m_EmuPaused;
 	bool		m_StartPaused;
 	bool		m_WasPaused;
+	bool		m_J2KWasEnabled;
```
I'm not sure I follow what this variable is being used for

monkeyman79 replied:
Joystick to keyboard mapping is temporarily enabled for displaying mapping dialog. This variable keeps previous state, to restore when the dialog is closed.

chrisn commented on Src/userkybd.cpp:
```diff
@@ -44,37 +45,244 @@ static void DrawBorder(HDC hDC, RECT rect, BOOL Depressed);
+// item in BBCKeys table, not a copy of it.
+int GetBBCKeyIndex(const BBCKey& key)
+{
+    int index = &key - BBCKeys;
```
I think this may be clearer if the function takes a `BBCKey` pointer instead of a reference, and the various map functions (`GetBBCKeyByResId`) could return pointers.

chrisn commented on Src/userkybd.cpp:
```diff
@@ -44,37 +45,244 @@ static void DrawBorder(HDC hDC, RECT rect, BOOL Depressed);
+    using resIdToKeyMapType = std::map<int, const BBCKey*>;
+
+    // Construct map on first use by lambda
+    static resIdToKeyMapType resIdToKeyMap = []()
```
Can this map (and the similar ones in other functions) be declared `const`?

monkeyman79 replied:
Definitely

richard-broadhurst commented:
I didn't see q mapping for the POV, which on my current favourite arcade stick (PS3 HORI Fighting stick mini) is where the digital joystick maps.

monkeyman79 commented:
@richard-broadhurst Hello :), It's there, named "hat" in the JoystickMap.txt, Axis7 and Axis8. Diagonals work too.
``` c++
	if (joyInfoEx.dwPOV != JOY_POVCENTERED)
	{
		if (joyInfoEx.dwPOV >= 29250 || joyInfoEx.dwPOV < 6750)
			axes |= (1 << BEEB_JOY_AX_HAT_UP);
		if (joyInfoEx.dwPOV >= 2250 && joyInfoEx.dwPOV < 15750)
			axes |= (1 << BEEB_JOY_AX_HAT_RIGHT);
		if (joyInfoEx.dwPOV >= 11250 && joyInfoEx.dwPOV < 24750)
			axes |= (1 << BEEB_JOY_AX_HAT_DOWN);
		if (joyInfoEx.dwPOV >= 20250 && joyInfoEx.dwPOV < 33750)
			axes |= (1 << BEEB_JOY_AX_HAT_LEFT);
	}
```

richard-broadhurst commented:
I'm really sorry, I missed that, next time I'll do it on the PC, not a phone!

chrisn commented:
Sorry for the delay. An interesting question was raised on Stardot about the joystick numbering. If we have different joysticks connected, is there a way to know which one maps to Joy1 or Joy2 in the config? Or I wonder if we should add UI to allow the user to easily remap them?

monkeyman79 commented:
Yeah, I was thinking about it too. It's just like Coeus said - windows will assign joystick number based on the order in which they appear in the system, so they can change between reboots, or just by disconnecting and reconnecting usb joysticks. I think that we could store joystick order in preferences e.g. like this:
```
JoystickOrder1=<mid1>:<pid1>
JoystickOrder2=<mid2>:<pid2>
JoystickOrder3=<mid3>:<pid3>
...
```
and provide simple UI to change the order. You just have to keep in mind that there may be several joysticks with the same manufacturer and product id connected.

Actually I'm thinking that maybe I should just go ahead and merge two-joysticks branch to this, because there is already part of the functionality there and otherwise I will just duplicate the job and will have to fix the conflicts later.

chrisn commented:
Yes, let's do that. It'll be easier to work on a single branch, I think.

monkeyman79 commented:
Summary of changes so far:

Second joystick and Joystick to keyboard mapping

* Assigning host joystick axes and buttons to BBC keyboard (up to 4 joysticks or gamepads separately).
* New human readable file format for joystick to keyboard mapping (see Documents/JoystickMap.txt)
* Changed joystick API from WinMM to DirectInput
* Display joystick names in menu
* Option to automatically load image-specific joystick mapping file (same as disk name, but with .jmap extension)
* Added "unassign" to keyboard and joystick mapping dialog
* Highlight assigned keys in joystick mapping dialog
* Added second BBC joystick support
* Select PC joystick and axes for each BBC joystick
* Save joystick order in preferences file
* Added "Change Joystick Order" dialog
* Added JoystickSensitivity parameter (only in preferences.cfg, no UI)
* Change Digital Mousestick handling

  This fixes an issue where the joystick position would never be centred. With this change, the joystick is centred when the mouse is in the centre of the BeebEm window.

* Map additional joystick buttons to analogue port PB0 and PB1
* Removed "Basic Hardware Only" option

  This option disabled the ADC and Serial interface, presumably to make the emulator run faster.

  This option was removed as it doesn't affect performance particularly, and simplifies joystick handling.

* Checkmark next to "Rescan Joysticks" menu item

  Display checkmark next to "Rescan Joysticks" menu item if first joystick initialised OK.

  Display warning message only if initialisation for "Rescan Joysticks" fails.

* Unmute when keyboard mapping dialog is closed
* Separate fire buttons on BBC joystick

  Map first two PC joystick or gamepad fire buttons to two BBC joystick buttons separately when only one BBC joystick is configured, . Fixes #74

monkeyman79 commented:
Moving to PR #95
